aws/resource_aws_cognito_identity_provider.go

```
@@ -0,0 +1,181 @@
+	conn := meta.(*AWSClient).cognitoidpconn
+	log.Print("[DEBUG] Creating Cognito Identity Provider")
+
+	name := aws.String(d.Get("provider_name").(string))
+	params := &cognitoidentityprovider.CreateIdentityProviderInput{
+		ProviderName: name,
+		ProviderType: aws.String(d.Get("provider_type").(string)),
+		UserPoolId:   aws.String(d.Get("user_pool_id").(string)),
+	}
+
+	if v, ok := d.GetOk("attribute_mapping"); ok {
+		params.AttributeMapping = expandCognitoIdentityProviderMap(v.(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("provider_details"); ok {
+		params.ProviderDetails = expandCognitoIdentityProviderMap(v.(map[string]interface{}))
+	}
+
+	if v, ok := d.GetOk("idp_identifiers"); ok {
+		params.IdpIdentifiers = expandStringList(v.([]interface{}))
+	}
+
+	_, err := conn.CreateIdentityProvider(params)
+	if err != nil {
+		return fmt.Errorf("Error creating Cognito Identity Provider: %s", err)
+	}
+
+	d.SetId(*name)
+
+	return resourceAwsCognitoIdentityProviderRead(d, meta)
+}
+
+func resourceAwsCognitoIdentityProviderRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).cognitoidpconn
+	log.Printf("[DEBUG] Reading Cognito Identity Provider: %s", d.Id())
+
+	ret, err := conn.DescribeIdentityProvider(&cognitoidentityprovider.DescribeIdentityProviderInput{
+		ProviderName: aws.String(d.Id()),
+		UserPoolId:   aws.String(d.Get("user_pool_id").(string)),
+	})
+
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	ip := ret.IdentityProvider
+	d.Set("provider_name", ip.ProviderName)
+	d.Set("provider_type", ip.ProviderType)
+	d.Set("user_pool_id", ip.UserPoolId)
+
+	if err := d.Set("attribute_mapping", flattenCognitoIdentityProviderMap(ip.AttributeMapping)); err != nil {
+		return fmt.Errorf("[DEBUG] Error setting attribute_mapping error: %#v", err)
```

Minor nitpick: these error messages will be returned to the operator, so they are not `DEBUG` and should return the normal error string via `%s`